### PR TITLE
CHJIT custom memory manager

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]
Ok. I've modified code a little, changed style, added attribution and a comment.

File: src/Interpreters/JIT/CHJIT.cpp [hunk 1]
```diff
@@ -113,25 +120,133 @@ class JITModuleMemoryManager
...
+            return std::error_code();
+        }
 
         size_t allocated_size = 0;
+
+    private:
+
+        static void protectBlock(const llvm::sys::MemoryBlock & block, int protection_flags)
+        {
+            int res = ::mprotect(block.base(), block.allocatedSize(), protection_flags);
```
Comment (azat) [inline src/Interpreters/JIT/CHJIT.cpp:209]: Now it is possible to get unaligned address:

```
$ strace -qq -f -e mmap,mprotect,munmap -p $(pgrep -f clickhouse.*serv)
[pid  6167] mprotect(0x7fff197fb000, 4096, PROT_READ|PROT_WRITE) = 0
[pid  6167] mprotect(0x7fff1982e000, 4096, PROT_READ|PROT_WRITE) = 0
[pid  6167] mprotect(0x7fff197fb000, 835, PROT_READ|PROT_EXEC) = 0
[pid  6167] mprotect(0x7fff1982e000, 8, PROT_READ) = 0
[pid  6167] mprotect(0x7fff1982e010, 16, PROT_READ) = -1 EINVAL (Invalid argument)
[pid  6167] mprotect(0x7fff197fb000, 4096, PROT_READ|PROT_WRITE) = 0
[pid  6167] mprotect(0x7fff1982e000, 4096, PROT_READ|PROT_WRITE) = 0
```

For the simple

```sql
select * from numbers(10000) where number+2>0 format Null
```

**Review by azat — COMMENTED** [review]
I'm not sure that it will fix the problem, how about adding a test something like https://github.com/azat/ClickHouse/commit/49d24d69ed6a6b0d2f0caec64d345a72f85b5ece

**Comment by azat** [comment]
@kitaisreal can you describe the purpose of the patch? AFAICS it does not address the VMA problem (#28156)

**Comment by kitaisreal** [comment]
@azat purpose of this patch is to use pages from underlying allocator (glibc malloc, jemalloc) and set protection bits to them without using mmap for each JITMemoryManger.
From my perspective this should fix problem #28156 that we have with mmap for each JITMemoryManager or not ?

**Comment by azat** [comment]
>@azat purpose of this patch is to use pages from underlying allocator (glibc malloc, jemalloc) and set protection bits to them without using mmap for each JITMemoryManger.

This will not help to avoid splitting of VMA, since changing protection bit is the problem.
You can take a look at my test (https://github.com/azat/ClickHouse/commit/49d24d69ed6a6b0d2f0caec64d345a72f85b5ece) and eventually it fails with `CANNOT_MPROTECT`.

I think the following (https://github.com/ClickHouse/ClickHouse/pull/28156#issuecomment-905809246 ) should help:

>The problem is that you need to change page protection bits, and this will likely split this range into separate VMA, another option is to create vma via mmap with a greater size and change protection bits for the whole range, then allocate memory from that range, this will fix the problem too.

JFYI Here is an example of split_vma call:

```
  - 50.00% 0x104e8b48068b4800
       0x2
       __mprotect
       entry_SYSCALL_64
       do_syscall_64
       __x64_sys_mprotect
       do_mprotect_pkey
       mprotect_fixup
       split_vma
```

**Comment by kitaisreal** [comment]
@azat agreed need to create special Allocator for CHJIT.

**Pull request merged**